### ISSUE TITLE
Incremental indexing (adding new content)

### ISSUE THREAD

**Issue by natoverse** [microsoft/graphrag#741]
### Do you need to file an issue?

- [ ] I have searched the existing issues and this feature is not already filed.
- [ ] My model is hosted on OpenAI or Azure. If not, please look at the "model providers" issue and don't file a new one here.
- [ ] I believe this is a legitimate feature request, not just a question. If this is a question, please use the Discussions area.

### Is your feature request related to a problem? Please describe.

A number of users have asked how to add new content to an existing index without needing to re-run the entire process. This is a feature we are planning, and are in the design stages now to ensure we have an efficient approach.

As it stands, new content can be added to a GraphRAG index without requiring a complete re-index. This is because we rely heavily on a cache to avoid repeating the same calls to the model API. There are several stages within the pipeline for which this is very efficient - namely those stages that are atomic and do not have upstream dependencies. For example, if you add new documents to a folder, we will not re-chunk existing documents or perform entity and relationship extraction on the existing content; we will simply fetch the processed content from the cache and pass it on. The new documents _will_ be processed and new entities and relationships extracted. Downstream of this, the graph construction process will need to recreate the graph to include the new nodes and edges, and communities will be recomputed - resulting in re-summarization, etc. You can get a sense of this process and what downstream steps may be re-processed by looking at the [data flow diagram](https://microsoft.github.io/graphrag/posts/index/1-default_dataflow/).

### Describe the solution you'd like

An ideal solution would be to add a new command to GraphRAG such as `update` that can be run against new data and augment an existing index. Considerations here include things such as evaluating the new entities to determine if they can be added to an existing community, and when those communities have been altered enough to constitute a "drift" that needs recomputing. We could also perform analysis to determine which communities have been edited, such that we ignore summarization on those that haven't changed.

### Additional context

We also need to consider the types of analysis incremental ingest can enable beyond just "updates". For example, daily ingest of news with thoughtful graph construction/annotation could allow for delta analysis such that questions like "what happened with person _x_ in the last 24 hours" or "catch me up on the news themes this week".

**Comment by natoverse** [comment]
This is a popular request, so I'm going to pin it and route other issues here.

**Comment by natoverse** [comment]
Related: removing existing content, e.g., #585

**Comment by KylinMountain** [comment]
can we split index into graph build and community summary? Lots of fans ask me if we can modify the knowledge graph in manual?

**Comment by KylinMountain** [comment]
If you modify the Ilm params in settings.yaml, all of cache will be invalid.

**Comment by natoverse** [comment]
Additional use case: adding files of a different type: https://github.com/microsoft/graphrag/issues/784

**Comment by ljhskyso** [comment]
Any ETA on this feature? Need this to assess whether I need to implement my own solution. @natoverse

**Comment by shaoqing404** [comment]
I think it is more urgent to change the cache from file to milvus(or more vector DB). The bottleneck that affects the overall query timeliness of graphrag has a significant impact on IO in relatively large files.

**Comment by vishyarjun** [comment]
there are two distinct scenarios for adding documents to your index, each requiring a different approach to community management and querying:
**_Scenario 1: Siloed Document Communities_**

- Concept: Each new document creates its own independent community within the index. This is like having separate folders, each with its own set of files.
- Community Management: Communities are created and maintained separately for each document. You'll need a mechanism to track which documents belong to which community.
- Querying: Queries are directed at a specific document community. The query pipeline only searches within that community, ensuring results are exclusively from the selected document.

**_Scenario 2: Unified Document Collection_**

- Concept: New documents are integrated into existing communities, enriching the existing knowledge base. This is like adding new files to a shared folder.
- Community Management: You manage communities at a higher level than individual documents. New content is tagged or categorized to associate it with the appropriate community.
- Querying: Queries search across all communities (or a subset you specify). Results may come from a mix of documents.

**Comment by gusye1234** [comment]
Hi, maybe checkout this [repo](https://github.com/gusye1234/nano-graphrag/tree/main), it supports incremental insert for entities and relationships. Also will compute the hash of the docs so only insert the new docs everytime you insert

**Comment by shandianshan** [comment]
@gusye1234 Hi, thank you for sharing your work. Could you please explain how you addressed the issue of integrating new entities into the community.

**Comment by gusye1234** [comment]
> @gusye1234 Hi, thank you for sharing your work. Could you please explain how you addressed the issue of integrating new entities into the community.

Sure. `nano-graphrag` use the md5 hash of docs and chunks as their key. When inserting begins, the same docs and chunks will be ignored and only the new chunks will continue to insert.
`nano-graphrag` will automatically load the previous graph from the working dir, and the new entities and relationships will be added on the current graph.

However, everytime you insert, `nano-graphrag` will still re-compute the communities and generate new community reports. The incremental update of communities is not yet implemented.

**Comment by Tipik1n** [comment]
The lack of this feature is also holding me back ( and probably many more ) from fully committing to using this repo as a solution in the gen ai space, Would love to see how you implement this.

**Comment by yaroslavyaroslav** [comment]
> Hi, maybe checkout this [repo](https://github.com/gusye1234/nano-graphrag/tree/main),

@gusye1234  If you changed this repo by deleting the branch specification, it would be accessible as the GitHub repo from within the GitHub mobile app, which I suppose would increase the number of stars by making it easier for mobile users. Because right now, it just opens the working directory instead.

**Comment by Einengutenmorgen** [comment]
I  really would like to be able to add documents to the index, as this is a key feature for a KG /KB.
In the original Blog Post it's often talked about the scalability of the Leiden Algorithm, but if the Framework is unable to provide efficient updating or fast reindexing, the scalability is useless. 


It took 2.5 h to index 4 software manuals and costed around 300 $ (OpenAi, Gpt4o-mini). As this was only half my data and most cost are caused by the community summaries.Therefor the current approach makes the project useless, except for tiny experiments. 
But even here it's limited. For example if I would like to test against truthfulQA (wikipedia only), it's still 395 wiki pages to index, what's properly killing all OpenAi rate limits.
Rereading the paper after testing:
"To combine the strengths of these contrasting methods, we propose a Graph RAG approach to question answering over private text corpora that scales with [..] the quantity of source text to be in- dexed." (From Local to Global, p1)
as the kids call it CAP

**Comment by JViggiani** [comment]
+1 to feature request. I'd need this feature to use graphrag as a solution to my problem space

**Comment by SZabolotnii** [comment]
+1 to feature request, it's a critical missing component.

**Comment by mellanon** [comment]
+1 to feature request. It was quite expensive to index my graph. I have one more document that needs to be added to the graph. A bit annoying but I'll leave it until this feature is available so that I don't end up triggering an expensive re-indexing exercise.

**Comment by Andrew-00** [comment]
I agree with everyone here. Incremental indexing is critical and important. Ideally also every time you query you can choose which of your indexed books/documents should be used, all of them or only a selection.

**Comment by KennyDizi** [comment]
+ 1

**Comment by wangiii** [comment]
+1

**Comment by zh-nj** [comment]
+1

**Comment by owquresh** [comment]
How can we handle cases where our data keeps changing over time? Not just added documents but changes being made to documents?

**Comment by aerwin-ds** [comment]
Are there any updates for when a solution for incremental indexing will be available in the official repo?

**Comment by AlonsoGuevara** [comment]
Folks,
In #1318 we merged our first version for Incremental indexing. I'm working on adding documentation, but those running from source are more than welcome to try it and report any issue.

**Comment by antoniocirclemind** [comment]
Hi, it might be worth mentioning this [repo](https://github.com/circlemind-ai/fast-graphrag) too. We just open-sourced it and it supports incremental updates.

**Comment by arastogi1111** [comment]
@AlonsoGuevara  The update command seems to be failing at creating the final documents. It takes the sweet time to index the new document(s) and update them to the existing graph and populates the `delta` folder but finally encounters a `KeyError: "['title'] not in index"` at :  `/usr/local/lib/python3.10/dist-packages/graphrag/index/update/incremental_index.py:281 in _update_entities`

**Comment by MarkusGutjahr** [comment]
> [@AlonsoGuevara](https://github.com/AlonsoGuevara) The update command seems to be failing at creating the final documents. It takes the sweet time to index the new document(s) and update them to the existing graph and populates the `delta` folder but finally encounters a `KeyError: "['title'] not in index"` at : `/usr/local/lib/python3.10/dist-packages/graphrag/index/update/incremental_index.py:281 in _update_entities`

also getting a similar error, after updating to the newest version of graphrag

**Comment by MarkusGutjahr** [comment]
> > [@AlonsoGuevara](https://github.com/AlonsoGuevara) The update command seems to be failing at creating the final documents. It takes the sweet time to index the new document(s) and update them to the existing graph and populates the `delta` folder but finally encounters a `KeyError: "['title'] not in index"` at : `/usr/local/lib/python3.10/dist-packages/graphrag/index/update/incremental_index.py:281 in _update_entities`
> 
> also getting a similar error, after updating to the newest version of graphrag

when i just run the indexing process, with no existing files, it gets an error for a missing file ("ValueError: Could not find create_final_documents.parquet in storage!")

when i use an exsting indexng result as the storage (as basedir in the settings.yaml), it starts the indexer process, but fails with: "KeyError: "['title'] not in index"

also it seems to make no difference between using "graphrag index" and "graphrag update", but just start te indexing with the update process


so right now, with the newest version, i cant update the existing files and also cant create new files (indexer generally not working)

**Comment by LuWei6896** [comment]
> > > [@AlonsoGuevara](https://github.com/AlonsoGuevara)更新命令似乎无法创建最终文档。它花费了大量时间来索引新文档并将其更新到现有图表并填充文件`delta`夹，但最终遇到了`KeyError: "['title'] not in index"`：`/usr/local/lib/python3.10/dist-packages/graphrag/index/update/incremental_index.py:281 in _update_entities`
> > 
> > 
> > 更新到最新版本的 graphrag 后，也出现了类似的错误
> 
> 当我刚刚运行索引过程时，由于没有现有文件，它会因缺少文件而出现错误（“ValueError：在存储中找不到 create_final_documents.parquet！”）
> 
> 当我使用现有的索引结果作为存储（作为 settings.yaml 中的 basedir）时，它会启动索引器进程，但会失败：“KeyError：“['title'] 不在索引中”
> 
> 而且使用“graphrag index”和“graphrag update”似乎没有区别，只是从更新过程开始索引
> 
> 所以现在，使用最新版本，我无法更新现有文件，也无法创建新文件（索引器通常不起作用）
> 
> 更新： 为了使用于创建新文件的正常索引器正常工作，需要删除 settings.yaml 中的“update_index_storage：”条目。 然后，当想要更新现有 gaph 数据时，需要添加它，同时将“storage：”的 basedir 更改为已创建的索引器结果。 这样我就可以让它工作了

  ![Image](https://github.com/user-attachments/assets/31de6847-7bf7-4a59-858d-de56a7249df9)
![Image](https://github.com/user-attachments/assets/558a3e61-00a9-4043-80db-12f12ff80c4f)

**Comment by LuWei6896** [comment]
> > > [@AlonsoGuevara](https://github.com/AlonsoGuevara) The update command seems to be failing at creating the final documents. It takes the sweet time to index the new document(s) and update them to the existing graph and populates the `delta` folder but finally encounters a `KeyError: "['title'] not in index"` at : `/usr/local/lib/python3.10/dist-packages/graphrag/index/update/incremental_index.py:281 in _update_entities`
> > 
> > 
> > also getting a similar error, after updating to the newest version of graphrag
> 
> when i just run the indexing process, with no existing files, it gets an error for a missing file ("ValueError: Could not find create_final_documents.parquet in storage!")
> 
> when i use an exsting indexng result as the storage (as basedir in the settings.yaml), it starts the indexer process, but fails with: "KeyError: "['title'] not in index"
> 
> also it seems to make no difference between using "graphrag index" and "graphrag update", but just start te indexing with the update process
> 
> so right now, with the newest version, i cant update the existing files and also cant create new files (indexer generally not working)
> 
> UPDATED: to make the normal indexer forcreaing new files work properly, the "update_index_storage:"-entry in the settings.yaml needs to be removed. then when wanting to update existing gaph data, it needs to be added with also changing the basedir of "storage:" to the alredy created indexer results. with this i get it to work

I solved the problem,u need to genarate a new settings.yaml，this error is related to the format and syntax errors of the configuration file.

**Comment by MarkusGutjahr** [comment]
> > > > [@AlonsoGuevara](https://github.com/AlonsoGuevara) The update command seems to be failing at creating the final documents. It takes the sweet time to index the new document(s) and update them to the existing graph and populates the `delta` folder but finally encounters a `KeyError: "['title'] not in index"` at : `/usr/local/lib/python3.10/dist-packages/graphrag/index/update/incremental_index.py:281 in _update_entities`
> > > 
> > > 
> > > also getting a similar error, after updating to the newest version of graphrag
> > 
> > 
> > when i just run the indexing process, with no existing files, it gets an error for a missing file ("ValueError: Could not find create_final_documents.parquet in storage!")
> > when i use an exsting indexng result as the storage (as basedir in the settings.yaml), it starts the indexer process, but fails with: "KeyError: "['title'] not in index"
> > also it seems to make no difference between using "graphrag index" and "graphrag update", but just start te indexing with the update process
> > so right now, with the newest version, i cant update the existing files and also cant create new files (indexer generally not working)
> > UPDATED: to make the normal indexer forcreaing new files work properly, the "update_index_storage:"-entry in the settings.yaml needs to be removed. then when wanting to update existing gaph data, it needs to be added with also changing the basedir of "storage:" to the alredy created indexer results. with this i get it to work
> 
> I solved the problem,u need to genarate a new settings.yaml，this error is related to the format and syntax errors of the configuration file.



yes, for testing i just added and removed the part related to "update_index_storage:". 
later i also created 2 seperate files for normal indexing and updating.

**Comment by radiant-tangent** [comment]
Since the scope for this issue is limited to adding documents, are there plans in the future to support updating the graph for documents that are removed? I see a lot of related issues got directed to this issue.

**Comment by ntsarb** [comment]
> Since the scope for this issue is limited to adding documents, are there plans in the future to support updating the graph for documents that are removed? I see a lot of issues related to removal got directed to this issue.

I will add that, in use cases where a new document may contain information that is meant to replace information from a previous document (e.g. a new law replacing a previous one), updating the knowledge graph and communities accordingly would be necessary.

**Comment by Desty27** [comment]
I explored the new incremental indexing  feature to add documents without rerunning the full pipeline. It has some issues.
During my test, the update flow failed with a KeyError: "['title'] not in index", triggered during the entities updation step. Maybe because settings.yaml is changed or stored Parquet files isn't correctly aligned.
It's a known limitation in the latest implementation, and others have reported similar issues in the GitHub thread. For now, a workaround is to re-run full indexing until the update command stabilizes.

**Comment by Mridul98** [comment]
Can anyone point me to the docs of incremental indexing please?

**Comment by natoverse** [comment]
We don't have any plans to invest futher in incremental updates to handle deletion or doc updates. Most folks have been fine removing docs and re-running regular indexed, relying on the cache to avoid repeat LLM API calls.
